### ISSUE TITLE
Duplicate instantiation of providers when provided via tokens

### ISSUE THREAD

**Issue by artakvg** [Urigo/graphql-modules#745]
I am trying to use my implementation for a provider with a token

```
providers: [
            { provide: DBProviderToken, useClass: AdminDBInitProvider },
...
]
```

But when we are using a provider with token some issues arise
-  at first it requires the AdminDBInitProvider to be provided in the module providers list. 
Simply said the one that is used for  _useClass_ to be provided in the _providers_ array also. 
- And by doing that it'll instantiate it twice. 

```
providers: [
            { provide: DBProviderToken, useClass: AdminDBInitProvider },
              AdminDBInitProvider
...
]
```

**Comment by ardatan** [comment]
@artakvg In the first one, when you inject `DBProviderToken` like below;
```
injector.get(DBProviderToken)` or `@Inject(DBProviderToken) dbInitProvider: AdminDbInitProvider`
```
but in the second one there are two different providers for DI;
`AdminDBInitProvider` class which uses the class itself as provider token and the other one I explained above. So the second one is not what you want as I understand.

```
providers: [
            { provide: DBProviderToken, useClass: AdminDBInitProvider },
...
]
```
While doing the first one, is there any error or something? If no, how do you try to inject this provider? Could you share a reproduction so I can help you better?
Thanks!

**Comment by artakvg** [comment]
create a simple module and put a class provider with token under providers
will throw en error I think. So tokens are not working with class providers.

**Comment by artakvg** [comment]
@ardatan here is a full example.

This is just a simple example following the documentation.

```
import 'reflect-metadata';
import { GraphQLModule, OnInit } from '@graphql-modules/core';
import { Inject, Injectable, ProviderScope } from '@graphql-modules/di';

const BProviderToken = Symbol.for('BProviderToken');

interface IBProvider {
    boo: () => void;
}

@Injectable({
    scope: ProviderScope.Application,
})
class AProvider implements OnInit {
    constructor(@Inject(BProviderToken) bProvider: IBProvider) {
        console.log('AProvider constructor');
    }

    public onInit() {
        console.log('AProvider onInit');
    }
}

class BProvider implements OnInit {
    constructor() {
        console.log('BProvider constructor');
    }

    public onInit() {
        console.log('BProvider onInit');
        this.boo();
    }

    public boo() {
        console.log('BProvider boo is called');
    }
}

const m = new GraphQLModule({
    providers:[ AProvider, {
        provide: BProviderToken, useClass: BProvider,
    }]
});

m.injector;
```

But without providing **BProvider** in provider list it'll log those 
```
BProvider constructor
AProvider constructor
AProvider onInit
```
and then errored 

```
Error: 
      GraphQL-Modules Error: Dependency Provider Not Found!
      - Provider #BProvider not provided in #Module ._8781164764539467 scope!

      Possible solutions:
      - Check if you have this provider in your module.
      - Check if you have the module of this provider imported in your dependent modules.
```

If I add the BProvider in the providers Array than it'll log this 
```
>BProvider constructor
>AProvider constructor
>AProvider onInit
>BProvider constructor // (why ????)
BProvider onInit
BProvider boo is called
BProvider onInit
BProvider boo is called
```

So in this case it'll create 2 instances and call onInit for them. 

The expected behavior as per documentation is that we should have just a single instance of the BProvider I guess. (as per application scope)

**Comment by artakvg** [comment]
Any updates on this @ardatan  ?

**Comment by alanlom** [comment]
Duplicates are also created with direct injection (without a token).
Checked with Apollo server (`apollo-server": "^2.9.5`) and graphql-modules (`@graphql-modules/core": "^0.7.14"`)
 
**Code to reproduced**
```
require ('reflect-metadata');
const { ApolloServer } = require('apollo-server');
import { GraphQLModule } from '@graphql-modules/core';
// import { configure, getLogger } from 'log4js';
import gql from 'graphql-tag';
import { Injectable, ProviderScope } from '@graphql-modules/di';
import { OnInit, OnRequest } from '@graphql-modules/core';

var globlNum = 0;

@Injectable({scope: ProviderScope.Application})
export class OnInitProvider implements OnInit, OnRequest  {
  constructor() {
    ++globlNum;
    console.log("constructor " + globlNum)
  }
  onInit() {
    console.info('OnInit ' + globlNum);
  }
  onRequest() {
    console.info('OnRequest ' + globlNum);
  }
}

const ModuleWithProvider = new GraphQLModule({
    providers: () => [
      OnInitProvider
    ]
});

const ModuleWithImport = new GraphQLModule({
  typeDefs: gql`
    type Query {
      getString: String
    }
  `,
  resolvers: {
    Query: {
      getString: () => "data" 
    }
  },
  imports: ( config ) => [
    ModuleWithProvider.forRoot(config)
  ],
});

const server = new ApolloServer({
  schema: ModuleWithImport.schema,
  context: session => session
});

server.listen().then(({ url }) => {
  console.log(`🚀  Server ready at ${url}`);
});
```

**Output results**
For the first request, we have duplicates:
```
Server ready at http://localhost:4000/
constructor 1
OnInit 1
ModuleWithProvider context
OnRequest 1
constructor 2
OnInit 2
ModuleWithImport context
```
The first instance is created, and all events are processed by it. Second instance is created for `ModuleWithImport`.  Only `OnInit` evet is processed by second.
For the next requests, only the second instance is used (the request is processed correctly):
```
ModuleWithProvider context
OnRequest 2
ModuleWithImport context
```

**Bug of processing OnRequest event**
Here we have other bug. See output results from above.
For the first request the `OnRequest` event is processed only by first provider instance.
There is no  `OnRequest` event for the second instance. 
As a result your first request failed when you need prepare something with `OnRequest` event.

I get such problems with *poolig example* from https://graphql-modules.com/docs/recipes/db-connection-pooling .
In the example provider contains event handler to create a new client in the `onRequest` method:
```
public onRequest() {
    this._poolClient = await this.pool.connect();
}
```
For the first request, the client will be created in `OnRequest` handler for first provider instance, but not for the second. As a result you get undefined client while process first request.
For the next query, everything works fine.

**Comment by kamilkisiela** [comment]
We released a new major version (v1.0) of GraphQL Modules. Here's the [migration guide](https://graphql-modules.com/docs/next/recipes/migration).

The new version solves the issue, so I'm closing the thread.

Please try to upgrade. If you can't and still want to use the legacy version, please open a new issue.
